docs(protocol): add inventory exchange (#115)
witnet/witnet-rust: docs/protocol/network/messages/inventory.md

The following diagram depicts the previous step under the assumption that the local node (`NodeA`) sends a `get_data` message requesting 3 blocks and 2 transactions.

```ascii
           NodeA                            NodeB
             +                                +
             |              INV               |
             +<-------------------------------+
             |                                |
             |           GET_DATA             |
             +------------------------------->+
             |                                |
             |             BLOCK              |
             +<-------------------------------+
             |             BLOCK              |
             +<-------------------------------+
             |             BLOCK              |
             +<-------------------------------+
             |              TX                |
             +<-------------------------------+
             |              TX                |
             +<-------------------------------+
             |                                |
             +                                +
```

## Get blocks message

> The `get_blocks` messages is used in order to notify the hash of the heighest known block by the peer. After the exchanging `get_blocks` messages between peers, the one with the longest blockchain in terms of blocks will send an `inv` message to the other peer. This message will include the list of blocks hashes starting right after the last known block hash provided by the other peer.

Reviewer: The `get_blocks` messages are used in order to notify the hash of the highest known block by the peer. After exchanging `get_blocks` messages between peers, the one with the longest blockchain in terms of blocks will send an `inv` message to the other peer. This message will include the list of block hashes starting right after the last known block hash provided by the other peer.

> The `get_blocks` message consists of the message header with the command `GET_BLOCKS` and as payload it contains the hash of the top (heighest) block known to the local peer:

Reviewer 2: ```suggestion
The `get_blocks` message consists of a message header with the `GET_BLOCKS` command and a payload containing the tip of the chain (the hash of the latest block) as known to the local peer:
```

| Field      | Type     | Description                                    |
| ---------- | :------: | ---------------------------------------------- |

> | last_block | [u32; 8] | hash of the last known block to the local node |

Reviewer 2: (``) and uppercase.

## Inv message

The `inv` message is used to advertise the knowledge of one or more objects (e.g. blocks, transactions, ...). The inventory message can be received unsolicited or in reply to a `get_blocks` message.

> The `inv` message consists of the message header with the command `INV` and as payload it contains a one more inventory entries:

Reviewer 2: ```suggestion
The `inv` message consists of a message header with the `INV` command and a payload containing one or more inventory entries:
```

| Field     | Type       | Description                 |
| --------- | :--------: | --------------------------- |

> | count     | u16        | number of inventory entries |

Reviewer 2: (``) and uppercase.

> | inventory | inv_vect[] | inventory vectors           |

Reviewer 2: (``) and uppercase.

> The type `inv_vect` (inventory vector) has the following data format:

Reviewer 2: ```suggestion
The `inv_vect` (inventory vector) data structure has the following schema:
```

| Field | Type     | Description                                  |
| ----- | :------: | -------------------------------------------- |

> | type  | u8       | type of object linked to the inventory entry |

Reviewer 2: (``) and uppercase.

> | count | [u32; 8] | hash of the object                           |

Reviewer 2: (``) and uppercase.

> The values of the type field are:

Reviewer 2: ```suggestion
The possible values for the `type` field are:
```

| Value | Name  | Description                          |
| ----- | :---: | ------------------------------------ |
| `0`   | ERROR | Data with this number may be ignored |
| `1`   | TX    | Hash is related to a transaction     |
| `2`   | BLOCK | Hash is related to a block           |

> 

Reviewer 2: ```suggestion
 | `3`   | `DATA_REQUEST` | Hash is related to a data request               | 
 | `4`   | `DATA_RESULT`  | Hash is related to the result of a data request | 
```

In the future, the type values may be extended in order to consider additional features.

## Get data messages

> The messages `get_data` are used in order to request specific objects from other nodes. Usually, `get_data` messages are sent after receiving a `inv` (inventory) message and filtering the already known objects. The response to a `get_data` message can be a `block` or `tx` message.

Reviewer 2: ```suggestion
The `get_data` messages are used in order to request specific objects from other nodes. Usually, `get_data` messages are sent after receiving a `inv` (inventory) message and filtering the already known objects. The response to a `get_data` message is often a message carrying the requested object (e.g.: `block`, `tx, etc.).
```

> The `get_data` message consists of the message header with the command `GET_DATA` and its payload has the following format:

Reviewer 2: ```suggestion
The `get_data` message consists of a message header with the `GET_DATA` command and a payload following this format:
```

| Field     | Type       | Description                 |
| --------- | :--------: | --------------------------- |

> | count     | u16        | number of inventory entries |

Reviewer 2: (``) and uppercase.

> | inventory | inv_vect[] | inventory vectors           |

Reviewer 2: (``) and uppercase.

## Block message

> The `block` message is used to transmit a single serialized block as response to a `get_data` message, which is requesting information given a block hash.

Reviewer: The `block` message is used to transmit a single serialized block as a response to a `get_data` message, which is requesting information given a block hash.

Reviewer: Not sure I understand "which is requesting information given a block hash."

> The `block` message consists of the message header with the command `BLOCK` and the payload is formatted as follows:

Reviewer 2: ```suggestion
The `block` message consists of a message header with the `BLOCK` command and a payload formatted as follows:
```

| Field     | Type         | Description                                                |
| --------- | :----------: | ---------------------------------------------------------- |

> | header    | block_header | block header, as described in the [Block] section          |

Reviewer 2: (``) and uppercase.

> | txn_count | u32          | total number of transactions in the block                  |

Reviewer 2: (``) and uppercase.

> | txns      | tx[]         | block transactions as described in the transaction section |

Reviewer 2: (``) and uppercase.

## Tx message

> Analogously, the `tx` message is used to transmit a single serialized transaction as response to a `get_data` message, which is requesting information given a transaction hash.

Reviewer: Analogously, the `tx` message is used to transmit a single serialized transaction as a response to a `get_data` message, which is requesting information given a transaction hash.

Reviewer: As above on line 142, I'm not sure I understand the use of "given" here.

> The `tx` message consists of the message header with the command `TX` and the payload contains a transaction information following the format defined in the transaction section:

Reviewer 2: ```suggestion
The `tx` message consists of a message header with the `TX` command and a payload containing information for a transaction following the format defined in the transaction section:
```

| Field | Type  | Description                                            |
| ----- | :---: | ------------------------------------------------------ |

> | tx    | tx    | transaction, as described in the [Transaction] section |

Reviewer 2: (``) and uppercase.
